diff --git a/src/backend/optimizer/util/pathnode.c b/src/backend/optimizer/util/pathnode.c
index fc97bf6..2ede28b 100644
--- a/src/backend/optimizer/util/pathnode.c
+++ b/src/backend/optimizer/util/pathnode.c
@@ -619,30 +619,13 @@ add_path(RelOptInfo *parent_rel, Path *new_path)
                }

                /*
-                * Remove current element from pathlist if dominated by new.
-                */
-               if (remove_old)
-               {
-                       parent_rel->pathlist = foreach_delete_current(parent_rel->pathlist,
-                                                                                                                 p1);
-
-                       /*
-                        * Delete the data pointed-to by the deleted cell, if possible
-                        */
-                       if (!IsA(old_path, IndexPath))
-                               pfree(old_path);
-               }
-               else
-               {
-                       /*
-                        * new belongs after this old path if it has more disabled nodes
-                        * or if it has the same number of nodes but a greater total cost
-                        */
-                       if (new_path->disabled_nodes > old_path->disabled_nodes ||
-                               (new_path->disabled_nodes == old_path->disabled_nodes &&
-                                new_path->total_cost >= old_path->total_cost))
-                               insert_at = foreach_current_index(p1) + 1;
-               }
+                * new belongs after this old path if it has more disabled nodes
+                * or if it has the same number of nodes but a greater total cost
+               */
+               if (new_path->disabled_nodes > old_path->disabled_nodes ||
+                       (new_path->disabled_nodes == old_path->disabled_nodes &&
+                               new_path->total_cost >= old_path->total_cost))
+                       insert_at = foreach_current_index(p1) + 1;

                /*
                 * If we found an old path that dominates new_path, we can quit
@@ -653,18 +636,8 @@ add_path(RelOptInfo *parent_rel, Path *new_path)
                        break;
        }

-       if (accept_new)
-       {
-               /* Accept the new path: insert it at proper place in pathlist */
-               parent_rel->pathlist =
-                       list_insert_nth(parent_rel->pathlist, insert_at, new_path);
-       }
-       else
-       {
-               /* Reject and recycle the new path */
-               if (!IsA(new_path, IndexPath))
-                       pfree(new_path);
-       }
+       /* Always accept the new path */
+       parent_rel->pathlist = list_insert_nth(parent_rel->pathlist, insert_at, new_path);
 }

 /*
@@ -869,21 +842,9 @@ add_partial_path(RelOptInfo *parent_rel, Path *new_path)
                        }
                }

-               /*
-                * Remove current element from partial_pathlist if dominated by new.
-                */
-               if (remove_old)
-               {
-                       parent_rel->partial_pathlist =
-                               foreach_delete_current(parent_rel->partial_pathlist, p1);
-                       pfree(old_path);
-               }
-               else
-               {
-                       /* new belongs after this old path if it has cost >= old's */
-                       if (new_path->total_cost >= old_path->total_cost)
-                               insert_at = foreach_current_index(p1) + 1;
-               }
+               /* Get location for new path if it has cost >= old */
+               if (new_path->total_cost >= old_path->total_cost)
+                       insert_at = foreach_current_index(p1) + 1;

                /*
                 * If we found an old path that dominates new_path, we can quit
@@ -894,17 +855,8 @@ add_partial_path(RelOptInfo *parent_rel, Path *new_path)
                        break;
        }

-       if (accept_new)
-       {
-               /* Accept the new path: insert it at proper place */
-               parent_rel->partial_pathlist =
-                       list_insert_nth(parent_rel->partial_pathlist, insert_at, new_path);
-       }
-       else
-       {
-               /* Reject and recycle the new path */
-               pfree(new_path);
-       }
+       /* Always accept the new path */
+       parent_rel->partial_pathlist = list_insert_nth(parent_rel->partial_pathlist, insert_at, new_path);
 }

 /*